src/NumberInput/Halogen/Component.purs

```diff
@@ -20,82 +20,100 @@ import Control.Monad.Except (runExcept)
 import Control.MonadPlus (guard)
 import DOM.Event.Event (Event)
 import Data.Bifunctor (lmap)
-import Data.Enum (class BoundedEnum, fromEnum, toEnum)
 import Data.Either (Either, either)
+import Data.Enum (class BoundedEnum, fromEnum, toEnum)
 import Data.Foreign (readBoolean, readString, toForeign)
 import Data.Foreign.Index (readProp)
 import Data.Int as Int
 import Data.Maybe (Maybe(..), fromMaybe, maybe)
 import Data.Number as N
 import Data.String (Pattern(..), length, stripSuffix)
-import Data.Tuple (Tuple(..), fst)
+import Data.Tuple (Tuple(..), fst, snd)
 import Halogen as H
 import Halogen.HTML as HH
 import Halogen.HTML.CSS as HCSS
 import Halogen.HTML.Core (ClassName)
 import Halogen.HTML.Events as HE
 import Halogen.HTML.Properties as HP
 import NumberInput.Range (Range(..), isInRange, rangeMax, rangeMin)
+import Unsafe.Reference (unsafeRefEq)
 
 
-type State val = InputValue val
+type State val =
+  { value ∷ InputValue val
+  , props ∷ Props val
+  }
 
 data Message val = NotifyChange (Input val)
 type Input val = Maybe val
 data Query val next
   = GetValue (Input val → next)
   | SetValue (Input val) next
   | Update (InputValue val) next
+  | SetProps (Props val) next
 
 type DSL val = H.ComponentDSL (State val) (Query val) (Message val)
 type HTML val = H.ComponentHTML (Query val)
 
-type Config val =
+type Props val =
   { title ∷ String
   , placeholder ∷ String
+  , hasNumberValue ∷ HasNumberInputValue val
   , range ∷ Range val
   , root ∷ Array ClassName
   , rootInvalid ∷ Array ClassName
-  , rootLength ∷ Int -> Array ClassName
+  , rootLength ∷ Int → Array ClassName
   }
 
+
 input ∷ ∀ val m
   . Ord val
-  ⇒ HasNumberInputVal val
-  → Config val
-  → H.Component HH.HTML (Query val) Unit (Message val) m
-input hasNumberInputVal conf = H.component
-  { initialState: const emptyNumberInputValue
-  , render: render hasNumberInputVal conf
-  , eval: eval hasNumberInputVal
-  , receiver: const Nothing
+  ⇒ H.Component HH.HTML (Query val) (Props val) (Message val) m
+input = H.component
+  { initialState: { value: emptyNumberInputValue, props: _}
+  , render: render
+  , eval: eval
+  , receiver: HE.input SetProps
   }
 
-render ∷ ∀ val. Ord val ⇒ HasNumberInputVal val → Config val → State val → HTML val
-render hasNumberInputVal conf num = numberElement hasNumberInputVal conf num
-
-eval ∷ ∀ val m . Eq val ⇒ HasNumberInputVal val → Query val ~> DSL val m
-eval hasNumberInputVal (SetValue number next) = do
-  H.put $ Tuple number (toMbString hasNumberInputVal number)
+eval ∷ ∀ val m . Eq val ⇒ Query val ~> DSL val m
+eval (SetProps props next) = do
+  state <- H.get
+  unless (unsafeRefEq props state.props) do
```

Comment:
Let's make this a proper equality test, there's no guarantee the parent is just passing the same record in every time it re-renders.